### DISCUSSION TITLE
Insert Pause

### DISCUSSION THREAD

**Discussion by korndog** [rhasspy/piper#199] [Q&A]
This TTS program sounds excellent!

Is there a way to insert a pause within the text string?

Thanks in advance!

-Vince

**Comment by korndog** [comment]
I guess there is no pause functionality. I tried extra comments and periods which apparently work for other TTS programs but no dice.  :(

**Reply by StoryHack** [reply]
I have to split the text where I want a pause, then generate separate files, then merge a "silence" wav in between.

**Reply by korndog** [reply]
@StoryHack Thanks for your response!  :)

**Reply by omega3** [reply]
@StoryHack Do you have a method to generate separate files from sentences in bulk? One file from one sentence. I also need a [method](https://github.com/rhasspy/piper/discussions/238) to apply filenames from csv to sentences.

**Reply by StoryHack** [reply]
I should probably rewrite it in python, but I scripted up something in php that chops up a text file by sentence, then builds a jsonl file, assigning filenames as it goes. At the same time it inserts those filenames into a "wavs-to-join" list. If there is ever a line in the text file that is just a tilde character "~", it doesn't make an entry for the jsonl, but inserts the filename for a silence.wav I made in to  the "wavs-to-join" list. After the script runs piper with the jsonl file, I use the "wavs-to-join" list to generate a command for the utility sox to actually join all the wavs in the correct order.

**Reply by StoryHack** [reply]
I don't have time or brainspace right this second to translate this to python, but here is a simplified version of what I am doing to generate from a longer .txt file. It is in php, but if you program at all, it should be easy enough to follow.

```
// note: I made a file in the same directory as this script 
//       called silence.wav. It is one second of silence at
//       a bit rate of 22050, which is the same I am generating
//       with Piper.

// get the text I need to generate
$text = file_get_contents("readme.txt");

// the final wav I want to produce
$out_file = "out.wav";

$json_cues = "";
$tmp_files = [];
$cue = 1; // I will increment this and use it as the temporary wav file names

// make a temporary directory for generated files
mkdir("./tmp");

// split up the text by paragraphs
$paragraphs = explode("\n", $text);

foreach($paragraphs as $paragraph){

	// get rid of any white space
	$paragraph = trim($paragraph);
	
	// if the paragraph is just an empty line, move to next paragraph without doing anything
	if($paragraph === ""){
		continue;
	}
	
	// if this paragraph is just a '~' character, insert a silence into the list of wavs to join,
	// then move on to next paragraph
	if($paragraph == "~"){
		$tmp_files[] = "silence";
		continue;
	}
	
	// now split paragraph into sentences, preserving punctuation marks
	$sentences = preg_split("/(?<=\?|\.|!)/", $paragraph);
	foreach($sentences as $sentence){
		// make the name of the wav to generate for this sentence
		$tmp_file = "./tmp/" . $cue . ".wav";
		
		// trim extra whitespace, sometime I get weird results from TTS systems with extra spaces...
		$sentence = trim($sentence);
		
		// if sentence is empty, skip to next
		if($sentence == ""){
			continue;
		}
		
		// escape any quotation marks (I've already made sure the file has no "slanty" ones, there are all straight quotes ")
		$sentence = str_replace('"', '\"', $sentence);
		
		// make the line I will need for the jsonl file
		$sentence_json = '{ "text": "' . $sentence . '", "output_file": "' . $tmp_file . '" }' . "\n";
		
		// add this cue to the ends of any other cues generated
		$json_cues .= $sentence_json;
		
		// add the file that will be generated to the list of wav files to join
		$tmp_files[] = $tmp_file;
		
		// increment the temp filename to use
		$cue++;	
	}
}

// write cues to file.
file_put_contents("./tmp/cues.jsonl", $json_cues);

// generate all wavs using the json file, assuming I have the proper files in the directory
$cmd = "cat ./tmp/cues.jsonl | ./piper --model en_US-lessac-medium.onnx --json-input ";
exec($cmd);

// merge the items in the wav files to join list to produce the final wav file
// note: sox must be installed on system
$cmd = "sox silence.wav " . implode(" ", $tmp_files) . " " . $out_file;
exec($cmd);

// cleanup temp files...
```

**Comment by SellersEvan** [comment]
Could we add support for a control character that would add a pause. Similar to the [AWS SSML TTS Tags](https://docs.aws.amazon.com/polly/latest/dg/supportedtags.html#break-tag)?

**Comment by Patrick-Ric** [comment]
Do you know that you can set the Pause time after every sentence with --sentence-silence?
Example: --sentence-silence 1.2 means 1.2 sec Pause time after every sentence.
Example:
./piper --model en_US-libritts_r-medium.onnx --speaker 111  --sentence-silence 1.2 --length-scale 1.5 --output_file ./sound.wav < ./textfile.txt

**Comment by JasonBlain** [comment]
I'm in Unity using the models and you can just append whatever amount of new float[] you want in the audio buffer after peeking at the model's samples output between each sentence return as per the Piper-Unity implementation available elsewhere on github here. I set it to a half second at whatever model sample rate. Works lovely.
